python/modules/IcePy/Util.cpp

```diff
@@ -223,15 +224,33 @@ IcePy::getFunction()
 
 IcePy::PyObjectHandle::PyObjectHandle(PyObject* p) : _p(p) {}
 
-IcePy::PyObjectHandle::PyObjectHandle(const PyObjectHandle& p) : _p(p._p) { Py_XINCREF(_p); }
+IcePy::PyObjectHandle::PyObjectHandle(const PyObjectHandle& p) : _p(p._p)
+{
+    AdoptThread adoptThread; // Ensure that we can call Python code.
+    Py_XINCREF(_p);
+}
 
-IcePy::PyObjectHandle::~PyObjectHandle() { Py_XDECREF(_p); }
+IcePy::PyObjectHandle::~PyObjectHandle()
+{
+    // The destructor can be called from Py_Exit at which point is no longer safe to call Python code.
+    // But this should only happen for objects that have been already released.
+
+    if (_p)
+    {
+        AdoptThread adoptThread; // Ensure that we can call Python code.
```

Comment:
Same here. We should understand who owns refcounts and who can release the last refcount.